infrastructure/src/main/java/org/corfudb/infrastructure/log/statetransfer/batchprocessor/committedbatchprocessor/CommittedBatchProcessor.java

```diff
@@ -0,0 +1,192 @@
+                LogUnitClient logUnitClientToTargetNode =
+                        runtimeLayout.getLogUnitClient(destinationNode);
+                LogUnitClient logUnitClientToCurrentNode =
+                        runtimeLayout.getLogUnitClient(currentNode);
+                try {
+                    ReadBatch readBatch = readRecords(addresses,
+                            Optional.of(destinationNode), logUnitClientToTargetNode);
+                    return writeRecords(readBatch, logUnitClientToCurrentNode,
+                            maxWriteRetries, writeSleepDuration);
+                } catch (ReadBatchException re) {
+                    log.warn("Read exception for node {} occurred. " +
+                            "Retry transfer for the next available node.", destinationNode, re);
+                } catch (Exception e) {
+                    log.error("Exception during batch transfer occurred. Failing the transfer.", e);
+                    throw e;
+                }
+            }
+            throw new ReadBatchException("No target nodes left to select from");
+        });
+    }
+
+    @Override
+    public CompletableFuture<TransferBatchResponse> transfer(
+            TransferBatchRequest transferBatchRequest) {
+        return tryTransferForRandomNodes(new RandomNodeIterator(transferBatchRequest))
+                .exceptionally(error -> TransferBatchResponse
+                        .builder()
+                        .transferBatchRequest(transferBatchRequest)
+                        .status(FAILED)
+                        .causeOfFailure(Optional.of(new StateTransferBatchProcessorException(
+                                "Failed batch: " + transferBatchRequest, error)))
+                        .build()
+                );
+    }
+
+    /**
+     * Read records directly from the randomly scheduled destination node (don't hole fill).
+     *
+     * @param addresses A batch of consecutive addresses.
+     * @param destNode  An optional destination node.
+     * @param client    A log unit client to the node.
+     * @return A read batch if the read went fine, an exception otherwise.
+     */
+    public ReadBatch readRecords(List<Long> addresses,
+                                 Optional<String> destNode,
+                                 LogUnitClient client) {
+        for (int i = 0; i < maxReadRetries; i++) {
+            try {
+                ReadResponse response = CFUtils
+                        .getUninterruptibly(client.readAll(addresses));
+                Map<Long, ILogData> records = new HashMap<>(response.getAddresses());
+                ReadBatch readBatch = checkReadRecords(addresses,
+                        records, destNode);
+                if (readBatch.getStatus() == ReadBatch.ReadStatus.FAILED) {
+                    throw new IllegalStateException("Some addresses failed to transfer: " +
+                            readBatch.getFailedAddresses());
+                }
```

Comment:
Is throw IllegalStateException here in order to retry read? Just want to confirm my understanding.